systemtest: AggShutdown wait 1s before `srv.Close`

marclop opened the pull request:
## Motivation/summary

It seems that `TestTransactionAggregationShutdown` is flaky when the
`srv.Close()` is called straightaway. Adding a 1 second sleep before
calling `srv.Close()`, seems to give enough time to the APM Server to
process the transaction and aggregate it.

## How to test these changes

Comment the `time.After` line and run:

```console
$ go test -count=20 -v -run TestTransactionAggregation .
```

I've seen the test fail as much as 2 times with a count of 10.

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-7067/detail/PR-7067/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-7067/detail/PR-7067/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-7067/detail/PR-7067/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-7067/detail/PR-7067/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_7067.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-01-13T07:32:17.446Z&rangeTo=2022-01-13T07:52:17.446Z&transactionName=apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=7a3b6326c6037c9746595937af200e36&transactionId=527d4b95d4cf518b)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-01-13T07:42:17.446+0000


* Duration: 45 min 22 sec


* Commit: a997083b75cace65a52db2dc581f48d17614088d



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 5624  |
| Skipped      | 20 |
| Total        | 5644   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/hey-apm` : Run the hey-apm benchmark.


- `/package` : Generate and publish the docker images.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

axw commented on systemtest/aggregation_test.go:
```diff
@@ -139,6 +139,11 @@ func TestTransactionAggregationShutdown(t *testing.T) {
 		estest.TermQuery{Field: "processor.event", Value: "transaction"},
 	)
 
+	// To avoid flaky tests, we wait for an additional 1s after the transaction
+	// has been indexed so the can aggregate the transaction before the server
+	// is stopped.
+	<-time.After(time.Second)
```
Shutting down the server should cause the aggregator to flush: https://github.com/elastic/apm-server/blob/06699bac0cefe16833238622a632b8aa667d7699/x-pack/apm-server/aggregation/txmetrics/aggregator.go#L151

Seems like a bug if that's not happening?

marclop commented:
Going to close this PR and create an issue to fix this behavior, so far I've been able to determine that it is because the top level context is cancelled after the metrics are added to the modelIndexer, but it fails to flush on `indexer.Close()`:

```
=== RUN   TestTransactionAggregationShutdown
    aggregation_test.go:162: context deadline exceeded
FAILED TEST!
Jan 14 11:28:40.865000 [debug]: bulk request completed: 1 indexed, 0 failed (0 exceeded capacity)
Jan 14 11:28:41.752000 [debug]: Cache size 1. Added ID systemtest_test.
Jan 14 11:28:41.752000 [info]: request ok
Jan 14 11:28:41.825000 [debug]: Received sigterm/sigint, stopping
Jan 14 11:28:41.825000 [info]: stopping apm-server... waiting maximum of 30 seconds for queues to drain
Jan 14 11:28:41.825000 [info]: Stop listening on: 127.0.0.1:0
Jan 14 11:28:41.825000 [debug]: no span metrics to publish
Jan 14 11:28:41.825000 [info]: service destinations aggregation stopped
Jan 14 11:28:41.825000 [debug]: publishing 1 metricsets
Jan 14 11:28:41.825000 [info]: transaction metrics aggregation stopped
Jan 14 11:28:41.826000 [info]: Server stopped
Jan 14 11:28:41.826000 [debug]: closing final batch processor
Jan 14 11:28:41.826000 [debug]: {Active:1 Added:2 BulkRequests:1 Failed:0 Indexed:1 TooManyRequests:0}
Jan 14 11:28:41.826000 [debug]: closed
Jan 14 11:28:41.826000 [debug]: {Active:1 Added:2 BulkRequests:1 Failed:0 Indexed:1 TooManyRequests:0}
Jan 14 11:28:41.826000 [error]: context canceled
Jan 14 11:28:41.826000 [debug]: bulkIndexeer.Flush(): ctx done, aborting flush
Jan 14 11:28:41.826000 [error]: bulk indexing request failed
Jan 14 11:28:41.827000 [info]: apm-server stopped.
    server.go:191: log file: /Users/marclop/repos/elastic/apm-server/systemtest/logs/TestTransactionAggregationShutdown/apm-server
--- FAIL: TestTransactionAggregationShutdown (61.27s)
```

A success run looks like this:

```
=== RUN   TestTransactionAggregationShutdown
Jan 14 11:30:30.667000 [debug]: bulk request completed: 1 indexed, 0 failed (0 exceeded capacity)
Jan 14 11:30:31.516000 [debug]: Cache size 1. Added ID systemtest_test.
Jan 14 11:30:31.516000 [info]: request ok
Jan 14 11:30:31.617000 [debug]: Received sigterm/sigint, stopping
Jan 14 11:30:31.617000 [info]: stopping apm-server... waiting maximum of 30 seconds for queues to drain
Jan 14 11:30:31.617000 [info]: Stop listening on: 127.0.0.1:0
Jan 14 11:30:31.617000 [debug]: no span metrics to publish
Jan 14 11:30:31.617000 [info]: service destinations aggregation stopped
Jan 14 11:30:31.618000 [debug]: publishing 1 metricsets
Jan 14 11:30:31.618000 [info]: transaction metrics aggregation stopped
Jan 14 11:30:31.619000 [info]: Server stopped
Jan 14 11:30:31.619000 [debug]: closing final batch processor
Jan 14 11:30:31.619000 [debug]: {Active:1 Added:2 BulkRequests:1 Failed:0 Indexed:1 TooManyRequests:0}
Jan 14 11:30:31.765000 [debug]: closed
Jan 14 11:30:31.765000 [debug]: {Active:0 Added:2 BulkRequests:2 Failed:0 Indexed:2 TooManyRequests:0}
Jan 14 11:30:31.765000 [error]: context canceled
Jan 14 11:30:31.765000 [info]: apm-server stopped.
--- PASS: TestTransactionAggregationShutdown (1.38s)
```
